initialize extended stack frame on RV32M1 ri5cy core

karstenkoenig opened the pull request:
The optional SOC_CONTEXT carries processor state registers that need to
be initialized properly to avoid uninitialized memory read as processor
state.
In particular on the RV32M1 the extra soc context stores a state for
special loop instructions, and loading non zero values will have the
core assume it is in a loop.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/26030

karstenkoenig committed:
soc: riscv: rv32m1: Fix optional context save

Saving an extended context for RV32M1 should be optional, but it was
broken due to the offset calculation not taking the according option
into account.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig committed:
arch: riscv: thread: Init soc context on stack

The optional SOC_CONTEXT carries processor state registers that need to
be initialized properly to avoid uninitialized memory read as processor
state.
In particular on the RV32M1 the extra soc context stores a state for
special loop instructions, and loading non zero values will have the
core assume it is in a loop.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig commented on soc/riscv/openisa_rv32m1/soc_context.h:
```diff
@@ -24,12 +24,12 @@
 
 /* Initial saved state. */
 #define SOC_ESF_INIT						\
```
I am unsure why this initially assigned `0xdeadband` here, it seems to have been the default context init value. But it wasn't really used to initialize the extended stack frame before. I tried to align it with the intention according to the documentation in https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/riscv/Kconfig#L51 .

karstenkoenig commented:
Rebased on master and checked again, still works :-)

MaureenHelm commented on soc/riscv/openisa_rv32m1/soc_offsets.h:
```diff
@@ -13,6 +13,7 @@
 
 #ifdef CONFIG_SOC_OPENISA_RV32M1_RI5CY
 
+#ifdef CONFIG_RISCV_SOC_CONTEXT_SAVE
```
This config is `default y if SOC_OPENISA_RV32M1_RI5CY`, so the extra ifdef seems redundant? Are you setting `CONFIG_SOC_OPENISA_RV32M1_RI5CY=n` explicitly?

karstenkoenig replied:
I understood it to be an option, defaulting to y, but also allowing n. So one can explicitly disable it if desired, as in less context switch overhead but not getting the benefit of the hardware loop optimization.

I did end up disabling it in a few intermediate configurations when trying to narrow the problem down, to figure out if it was a compiler issue with the hardware loops or the loop registers getting messed up during context switching and it helped there.

Personally I am not attached to having it as an option, just felt it was broken as is thus added this somewhat ugly additional ifdef level

karstenkoenig replied:
Guess I didn't fully answer the question. After fixing the root cause by zero initializing the loop registers in the stack frame I do not set `CONFIG_SOC_OPENISA_RV32M1_RI5CY=n` explicitly :-)

karstenkoenig replied:
@MaureenHelm Sorry I just reread this, i totally referenced the wrong config symbol.
The issue at hand is not `CONFIG_SOC_OPENISA_RV32M1_RI5CY`, but `CONFIG_RISCV_SOC_CONTEXT_SAVE`. It is `default y`, but can be configured to false if desired, to avoid the overhead of the stack operations if one doesn't use the loop optimization feature of the RI5CY core. This commit fixes `CONFIG_RISCV_SOC_CONTEXT_SAVE=n`. I can remove the change if you feel it doesn't need to be supported.

karstenkoenig committed:
tests: drivers: spi: spi_loopback: remove workaround

Due to issue #20589 DEBUG_OPTIMIZATIONS were enabled for the rv32m1 vega
ri5cy vega config. This is now resolved after making sure the spi tx
buffer is always initialized and the rv32m1 ri5cy core stack properly
initialized, so this workaround can be removed.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

karstenkoenig commented:
Hi, not sure how to have this proceed further. The PR mostly tagged RISC-V maintainers, but should be fairly understandable to other platforms as well, as it is just missing initialization on part of the stack frame that is required. @MaureenHelm do you know if there is anybody else I could tag on this? RV32M1 is sort of a niche thing, but still a very interesting platform

MaureenHelm commented:
> Hi, not sure how to have this proceed further. The PR mostly tagged RISC-V maintainers, but should be fairly understandable to other platforms as well, as it is just missing initialization on part of the stack frame that is required. @MaureenHelm do you know if there is anybody else I could tag on this? RV32M1 is sort of a niche thing, but still a very interesting platform

Sorry for the delay. These changes look ok to me, but would like one of the RISC-V maintainers to review.

@kgugala @pgielda @nategraff-sifive please review.

sjg20 commented on soc/riscv/openisa_rv32m1/soc_context.h:
```diff
@@ -24,12 +24,12 @@
-	0xdeadbaad,						\
-	0xdeadbaad,						\
-	0xdeadbaad
+	0,						\
```
If they are all zero, do you even need this #define?

karstenkoenig replied:
Not really, in the Kconfig it is documented as mandatory though
```
config RISCV_SOC_CONTEXT_SAVE
...
	  Enabling this option requires that the SoC provide a
	  soc_context.h header which defines the following macros:
...
	  - SOC_ESF_INIT: structure contents initializer for struct soc_esf
	    state. The last initialized member should not end in a comma.
```
Right now RV32M1_RI5CY is the only soc actually using this feature, so SOC_ESF_INIT could just be removed and it could be hardcoded to zero initialize. If any other core were to use the RISCV_SOC_CONTEXT_SAVE feature a static constant initializer might not fit all use cases anyways, as the init values on the stack could be runtime dependent. Should I just go ahead and replace it with a memset to 0?
